Code-4-Community/speak-for-the-trees-backend-v2
# Ot.rate limit v2 (Bucket4j server side throttling)

## Why

Resolves #<ticket number here>

Implements server-side IP-specific token bucket API throttling to prevent server overload from massive popularity or DOS attack.

## This PR

IpThrottlingFilter class instance acts as a VertX handler and intercepts all API calls before they get forwarded to processors. Keeps a hash map of IP addresses to buckets. Adds new buckets if new IP address. Uses bucket4j library.

Currently there is only one throttling class instance, so it is a globally-aggregated token consumption per IP. If we wanted to throttle per router or per route, we would create separate filter class instances for each route(r).

These are subject to change, but initial capacity is 20 and refills tokens at 10/min greedily (1 token every 6 sec) . One API route call consumes one token.

Service main contains these constants, creates an IpThrottlingFilter instance with the config (with the constants) and passes it to ApiRouter. The instance gets mounted to the common and protected router. If we wanted to to lower route(r) level throttling, we would create individual instances for each route(r) during subrouter initialization or instantiation.

## Verification Steps

Postman spamming to ensure that you get 429 error. Localhost reload spamming to get 429 error. Frontend might want to display different message when 429 occurs (Too many requests)?

<img width="972" alt="Screenshot 2024-07-19 at 4 39 05 PM" src="https://github.com/user-attachments/assets/d3f90c34-b4f0-4c1b-bb20-571051b84134">
<img width="1111" alt="Screenshot 2024-07-19 at 4 39 56 PM" src="https://github.com/user-attachments/assets/b8dd7a6a-ca59-4e80-8303-178db3f6d329">
